packages/calypso-router/src/index.js

```diff
@@ -0,0 +1,1207 @@
+ */
+function escapeRegExp( s ) {
+	return s.replace( /([.+*?=^!:${}()[\]|/\\])/g, '\\$1' );
+}
+
+/**
+ * Initialize a new "request" `Context`
+ * with the given `path` and optional initial `state`.
+ * @class
+ * @param {string} path
+ * @param {Object=} state
+ */
+function Context( path, state, pageInstance ) {
+	this.page = pageInstance;
+	const window = this.page._window;
+	const hashbang = this.page._hashbang;
+
+	const pageBase = this.page._getBase();
+	if ( '/' === path[ 0 ] && 0 !== path.indexOf( pageBase ) ) {
+		path = pageBase + ( hashbang ? '#!' : '' ) + path;
+	}
+	const i = path.indexOf( '?' );
+
+	this.canonicalPath = path;
+	const re = new RegExp( '^' + escapeRegExp( pageBase ) );
+	this.path = path.replace( re, '' ) || '/';
+	if ( hashbang ) {
+		this.path = this.path.replace( '#!', '' ) || '/';
+	}
+
+	this.title = hasDocument && window.document.title;
+	this.state = state || {};
+	this.state.path = path;
+	this.querystring = ~i ? this.page._decodeURLEncodedURIComponent( path.slice( i + 1 ) ) : '';
+	this.pathname = this.page._decodeURLEncodedURIComponent( ~i ? path.slice( 0, i ) : path );
+	this.params = {};
+
+	// fragment
+	this.hash = '';
+	if ( ! hashbang ) {
+		if ( ! ~this.path.indexOf( '#' ) ) {
+			return;
+		}
+		const parts = this.path.split( '#' );
+		this.path = this.pathname = parts[ 0 ];
+		this.hash = this.page._decodeURLEncodedURIComponent( parts[ 1 ] ) || '';
+		this.querystring = this.querystring.split( '#' )[ 0 ];
+	}
+}
+
+/**
+ * Push state.
+ */
+Context.prototype.pushState = function () {
+	this.page.len++;
+	if ( hasHistory ) {
+		this.page._window.history.pushState(
+			this.state,
+			this.title,
+			this.page._hashbang && this.path !== '/' ? '#!' + this.path : this.canonicalPath
+		);
+	}
+};
+
+/**
+ * Save the context state.
+ */
+Context.prototype.save = function () {
+	if ( hasHistory ) {
+		this.page._window.history.replaceState(
+			this.state,
+			this.title,
+			this.page._hashbang && this.path !== '/' ? '#!' + this.path : this.canonicalPath
+		);
+	}
+};
+
+/**
+ * Initialize `Route` with the given HTTP `path`,
+ * and an array of `callbacks` and `options`.
+ *
+ * Options:
+ * - `sensitive`    enable case-sensitive routes
+ * - `strict`       enable strict matching for trailing slashes
+ *
+ * @class
+ * @param {string} path
+ * @param {Object=} options
+ */
+function Route( path, options, pageInstance ) {
+	this.page = pageInstance;
+	const opts = options || {};
+	opts.strict = opts.strict || this.page._strict;
+	this.path = path === '*' ? '(.*)' : path;
+	this.method = 'GET';
+	this.keys = [];
+	this.regexp = pathToRegexp( this.path, this.keys, opts );
+}
+
+/**
+ * Return route middleware with
+ * the given callback `fn()`.
+ * @param {Function} fn
+ * @returns {Function}
+ */
+Route.prototype.middleware = function ( fn ) {
+	const handler = ( ctx, next ) => fn( ctx, next );
+	handler.match = ( ctx ) => {
+		if ( this.match( ctx.path, ctx.params ) ) {
+			ctx.routePath = this.path;
```

Comment:
Why is this line necessary? Did we use this for debugging at some point and forgot to clean it up?

Reply:
This `ctx.routePath` was in the original `Route.prototype.middleware` code, and I left it as is, just moved it around during the refactoring/bugfixing.

I think Calypso doesn't use `routePath` at all, in a followup we can remove it as a redundant feature.